reconcile/vault_replication.py

```diff
@@ -0,0 +1,392 @@
+    Vault.
+    * vault_instance - VaultInstanceV1 object from AppInterface Data.
+    """
+    vault_creds = {}
+    vault = cast(_VaultClient, VaultClient())
+
+    if not isinstance(
+        vault_auth,
+        VaultReplicationConfigV1_VaultInstanceAuthV1_VaultInstanceAuthApproleV1,
+    ) and not isinstance(
+        vault_auth,
+        VaultInstanceV1_VaultReplicationConfigV1_VaultInstanceAuthV1_VaultInstanceAuthApproleV1,
+    ):
+        raise VaultInvalidAuthMethod
+
+    role_id = {
+        "path": vault_auth.role_id.path,
+        "field": vault_auth.role_id.field,
+    }
+    secret_id = {
+        "path": vault_auth.secret_id.path,
+        "field": vault_auth.secret_id.field,
+    }
+
+    vault_creds["role_id"] = vault.read(role_id)
+    vault_creds["secret_id"] = vault.read(secret_id)
+    vault_creds["server"] = vault_address
+
+    return vault_creds
+
+
+def replicate_paths(
+    dry_run: bool,
+    source_vault: _VaultClient,
+    dest_vault: _VaultClient,
+    replications: VaultReplicationConfigV1,
+) -> None:
+
+    if replications.paths is None:
+        return
+
+    for path in replications.paths:
+
+        if isinstance(path, VaultReplicationJenkinsV1):
+            if path.policy is not None:
+                vault_query_data = vault_policies.query(query_func=gql.get_api().query)
+                policy_paths = get_policy_paths(
+                    path.policy.name,
+                    path.policy.instance.name,
+                    vault_query_data,
+                )
+            else:
+                policy_paths = None
+
+            jenkins_query_data = jenkins_configs.query(query_func=gql.get_api().query)
+            path_list = get_jenkins_secret_list(
+                source_vault, path.jenkins_instance.name, jenkins_query_data
+            )
+            check_invalid_paths(path_list, policy_paths)
+            for vault_path in path_list:
+                copy_vault_secret(dry_run, source_vault, dest_vault, vault_path)
+
+
+def get_start_end_secret(path: str) -> tuple[str, str]:
+
+    start = path[0 : path.index("{")]
+    if start[-1] != "/":
+        start = start.rsplit("/", 1)[0] + "/"
+    try:
+        end = path[path[path.index("}") : :].index("/") + path.index("}") : :]
+    except ValueError:
+        end = ""
+
+    return start, end
+
+
+def get_secrets_from_templated_path(
+    vault_instance: _VaultClient, path: str
+) -> list[str]:
+    """Returns a list of secrets that result from a templated path
+    expansion.
+
+    The input secret is the name of a jenkins instance to filter
+    the secrets:
+    * path - Vault path
+    """
+
+    secret_list = []
+    path_slices = path.split("/")
+    for s in path_slices:
+        if "{" in s and "}" in s:
+            template = s
+
+    cap_groups = re.search(r"(.*)(\{.*\})(.*)", template)
+    if cap_groups is not None:
+        prefix = cap_groups.group(1)
+        suffix = cap_groups.group(3)
+    else:
+        raise VaultInvalidPaths
```

Comment:
Please add nice and helpful comments to every exception you raise.

Reply:
Done!